Fatal error of json decode

I added `Stream: true` to the request parameters and returned some "Server-Sent Events" data that could not be parsed.

* Request: 
```go
client := chat.NewClient(s, "gpt-3.5-turbo")
resp, err := client.CreateCompletion(ctx, &chat.CreateCompletionParams{
    Messages: []*chat.Message{
	    {Role: "user", Content: "hello"},
    },
    Stream: true,
})
```
* Response:
```text

data: {"id":"chatcmpl-6s9Fmuz4z...jxmcQKQeiGr","object":"chat.completion.chunk","created":1678363302,"model":"gpt-3.5-turbo-0301","choices":[{"delta":{"role":"assistant"},"index":0,"finish_reason":null}]}
data: {"id":"chatcmpl-6s9Fmuz4z...jxmcQKQeiGr","object":"chat.completion.chunk","created":1678363302,"model":"gpt-3.5-turbo-0301","choices":[{"delta":{"content":"\n\n"},"index":0,"finish_reason":null}]}
data: {"id":"chatcmpl-6s9Fmuz4z...jxmcQKQeiGr","object":"chat.completion.chunk","created":1678363302,"model":"gpt-3.5-turbo-0301","choices":[{"delta":{"content":"Hello"},"index":0,"finish_reason":null}]}
data: {"id":"chatcmpl-6s9Fmuz4z...jxmcQKQeiGr","object":"chat.completion.chunk","created":1678363302,"model":"gpt-3.5-turbo-0301","choices":[{"delta":{"content":"!"},"index":0,"finish_reason":null}]}
data: {"id":"chatcmpl-6s9Fmuz4z...jxmcQKQeiGr","object":"chat.completion.chunk","created":1678363302,"model":"gpt-3.5-turbo-0301","choices":[{"delta":{"content":" How"},"index":0,"finish_reason":null}]}
data: {"id":"chatcmpl-6s9Fmuz4z...jxmcQKQeiGr","object":"chat.completion.chunk","created":1678363302,"model":"gpt-3.5-turbo-0301","choices":[{"delta":{"content":" can"},"index":0,"finish_reason":null}]}
data: {"id":"chatcmpl-6s9Fmuz4z...jxmcQKQeiGr","object":"chat.completion.chunk","created":1678363302,"model":"gpt-3.5-turbo-0301","choices":[{"delta":{"content":" I"},"index":0,"finish_reason":null}]}
data: {"id":"chatcmpl-6s9Fmuz4z...jxmcQKQeiGr","object":"chat.completion.chunk","created":1678363302,"model":"gpt-3.5-turbo-0301","choices":[{"delta":{"content":" assist"},"index":0,"finish_reason":null}]}
data: {"id":"chatcmpl-6s9Fmuz4z...jxmcQKQeiGr","object":"chat.completion.chunk","created":1678363302,"model":"gpt-3.5-turbo-0301","choices":[{"delta":{"content":" you"},"index":0,"finish_reason":null}]}
data: {"id":"chatcmpl-6s9Fmuz4z...jxmcQKQeiGr","object":"chat.completion.chunk","created":1678363302,"model":"gpt-3.5-turbo-0301","choices":[{"delta":{"content":" today"},"index":0,"finish_reason":null}]}
data: {"id":"chatcmpl-6s9Fmuz4z...jxmcQKQeiGr","object":"chat.completion.chunk","created":1678363302,"model":"gpt-3.5-turbo-0301","choices":[{"delta":{"content":"?"},"index":0,"finish_reason":null}]}
data: {"id":"chatcmpl-6s9Fmuz4z...jxmcQKQeiGr","object":"chat.completion.chunk","created":1678363302,"model":"gpt-3.5-turbo-0301","choices":[{"delta":{},"index":0,"finish_reason":"stop"}]}
data: [DONE]

Failed to complete: EOF
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x631bb6]
```

--- a/chat/chat.go
+++ b/chat/chat.go
@@ -3,9 +3,12 @@ package chat
 
 import (
 	"context"
+	"errors"
 
 	"github.com/rakyll/openai-go"
 )
+
+const defaultModel = "gpt-3.5-turbo"
 
 const defaultCreateCompletionsEndpoint = "https://api.openai.com/v1/chat/completions"
 
@@ -23,7 +26,7 @@ type Client struct {
 // and defaults to the given model.
 func NewClient(session *openai.Session, model string) *Client {
 	if model == "" {
-		model = "gpt-3.5-turbo"
+		model = defaultModel
 	}
 	return &Client{
 		s:                        session,
@@ -74,6 +77,9 @@ type Message struct {
 func (c *Client) CreateCompletion(ctx context.Context, p *CreateCompletionParams) (*CreateCompletionResponse, error) {
 	if p.Model == "" {
 		p.Model = c.model
+	}
+	if p.Stream {
+		return nil, errors.New("use StreamingClient instead")
 	}
 
 	var r CreateCompletionResponse

--- a/chat/streaming.go
+++ b/chat/streaming.go
@@ -1,0 +1,56 @@
+package chat
+
+import (
+	"context"
+
+	"github.com/rakyll/openai-go"
+)
+
+// StreamingClient is a client to communicate with Open AI's ChatGPT APIs.
+type StreamingClient struct {
+	s     *openai.Session
+	model string
+
+	// CreateCompletionsEndpoint allows overriding the default API endpoint.
+	// Set this field before using the client.
+	CreateCompletionEndpoint string
+}
+
+// NewStreamingClient creates a new default streaming client that uses the given session
+// and defaults to the given model.
+func NewStreamingClient(session *openai.Session, model string) *StreamingClient {
+	if model == "" {
+		model = defaultModel
+	}
+	return &StreamingClient{
+		s:                        session,
+		model:                    model,
+		CreateCompletionEndpoint: defaultCreateCompletionsEndpoint,
+	}
+}
+
+type CreateCompletionStreamingResponse struct {
+	ID        string             `json:"id,omitempty"`
+	Object    string             `json:"object,omitempty"`
+	CreatedAt int64              `json:"created_at,omitempty"`
+	Choices   []*StreamingChoice `json:"choices,omitempty"`
+}
+
+type StreamingChoice struct {
+	Delta        *Message `json:"delta,omitempty"`
+	Index        int      `json:"index,omitempty"`
+	LogProbs     int      `json:"logprobs,omitempty"`
+	FinishReason string   `json:"finish_reason,omitempty"`
+}
+
+func (c *StreamingClient) CreateCompletion(ctx context.Context, p *CreateCompletionParams, fn func(r *CreateCompletionStreamingResponse)) error {
+	if p.Model == "" {
+		p.Model = c.model
+	}
+	p.Stream = true
+
+	var r CreateCompletionStreamingResponse
+	return c.s.MakeStreamingRequest(ctx, c.CreateCompletionEndpoint, p, &r, func(r any) {
+		fn(r.(*CreateCompletionStreamingResponse))
+	})
+}

--- a/completion/completion.go
+++ b/completion/completion.go
@@ -3,6 +3,7 @@ package completion
 
 import (
 	"context"
+	"errors"
 
 	"github.com/rakyll/openai-go"
 )
@@ -77,6 +78,9 @@ type Choice struct {
 func (c *Client) Create(ctx context.Context, p *CreateParams) (*CreateResponse, error) {
 	if p.Model == "" {
 		p.Model = c.model
+	}
+	if p.Stream {
+		return nil, errors.New("use StreamingClient instead")
 	}
 
 	var r CreateResponse

--- a/completion/streaming.go
+++ b/completion/streaming.go
@@ -1,0 +1,40 @@
+package completion
+
+import (
+	"context"
+
+	"github.com/rakyll/openai-go"
+)
+
+// StreamingClient is a client to communicate with Open AI's completions API.
+type StreamingClient struct {
+	s     *openai.Session
+	model string
+
+	// CreateEndpoint allows overriding the default API endpoint.
+	// Set this field before using the client.
+	CreateEndpoint string
+}
+
+// NewStreamingClient creates a new default streaming client that uses the given session
+// and defaults to the given model.
+func NewStreamingClient(session *openai.Session, model string) *StreamingClient {
+	return &StreamingClient{
+		s:              session,
+		model:          model,
+		CreateEndpoint: defaultCreateEndpoint,
+	}
+}
+
+// Create creates a completion for the provided parameters.
+func (c *StreamingClient) Create(ctx context.Context, p *CreateParams, fn func(r *CreateResponse)) error {
+	if p.Model == "" {
+		p.Model = c.model
+	}
+	p.Stream = true
+
+	var r CreateResponse
+	return c.s.MakeStreamingRequest(ctx, c.CreateEndpoint, p, &r, func(r any) {
+		fn(r.(*CreateResponse))
+	})
+}

--- a/openai.go
+++ b/openai.go
@@ -2,6 +2,7 @@
 package openai
 
 import (
+	"bufio"
 	"bytes"
 	"context"
 	"encoding/json"
@@ -10,6 +11,7 @@ import (
 	"mime/multipart"
 	"net/http"
 	"net/url"
+	"strings"
 	"time"
 )
 
@@ -54,7 +56,52 @@ func (s *Session) MakeRequest(ctx context.Context, endpoint string, input, outpu
 		return err
 	}
 
-	return s.sendRequest(req, "application/json", output)
+	body, err := s.sendRequest(req, "application/json")
+	if err != nil {
+		return err
+	}
+	defer body.Close()
+
+	return json.NewDecoder(body).Decode(output)
+}
+
+func (s *Session) MakeStreamingRequest(ctx context.Context, endpoint string, input interface{}, output interface{}, fn func(any)) error {
+	const (
+		streamPrefix = "data: "
+		streamEnd    = "[DONE]"
+	)
+
+	buf, err := json.Marshal(input)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewReader(buf))
+	if err != nil {
+		return err
+	}
+
+	body, err := s.sendRequest(req, "application/json")
+	if err != nil {
+		return err
+	}
+	defer body.Close()
+
+	scanner := bufio.NewScanner(body)
+	for scanner.Scan() {
+		line := strings.Replace(scanner.Text(), streamPrefix, "", 1)
+		if line == "" {
+			continue
+		}
+		if line == streamEnd {
+			return nil
+		}
+		if err := json.Unmarshal([]byte(line), output); err != nil {
+			return fmt.Errorf("failed to unmarshal streaming response: %w", err)
+		}
+		fn(output)
+	}
+	return scanner.Err()
 }
 
 // Upload makes a multi-part form data upload them with
@@ -71,10 +118,16 @@ func (s *Session) Upload(ctx context.Context, endpoint string, file io.Reader, f
 	if err != nil {
 		return err
 	}
-	return s.sendRequest(req, mw.FormDataContentType(), output)
+	body, err := s.sendRequest(req, mw.FormDataContentType())
+	if err != nil {
+		return err
+	}
+	defer body.Close()
+
+	return json.NewDecoder(body).Decode(output)
 }
 
-func (s *Session) sendRequest(req *http.Request, contentType string, output any) error {
+func (s *Session) sendRequest(req *http.Request, contentType string) (io.ReadCloser, error) {
 	if s.apiKey != "" {
 		req.Header.Set("Authorization", "Bearer "+s.apiKey)
 	}
@@ -86,21 +139,20 @@ func (s *Session) sendRequest(req *http.Request, contentType string, output any)
 
 	resp, err := s.HTTPClient.Do(req)
 	if err != nil {
-		return err
+		return nil, fmt.Errorf("error making request: %w", err)
 	}
-	defer resp.Body.Close()
 
 	if resp.StatusCode < 200 || resp.StatusCode >= 400 {
 		respBody, err := io.ReadAll(resp.Body)
 		if err != nil {
-			return err
+			return nil, err
 		}
-		return &APIError{
+		return nil, &APIError{
 			StatusCode: resp.StatusCode,
 			Payload:    respBody,
 		}
 	}
-	return json.NewDecoder(resp.Body).Decode(output)
+	return resp.Body, nil
 }
 
 func upload(mw *multipart.Writer, file io.Reader, fileExt string, params url.Values) error {

